server/models/validations/notification.js

```diff
@@ -0,0 +1,32 @@
+const __ = require('config').universalPath
+const _ = __.require('builders', 'utils')
+const { pass, passFromFunction } = require('./common')
+const { types } = require('../attributes/notification')
+const Group = require('../group')
+
+const validations = module.exports = {
+  pass,
+  type: type => types.includes(type),
+  data: (data, { type }) => {
+    passFromFunction('data', data, _.isPlainObject)
+    dataValidationPerType[type](data)
+    return true
+  }
+}
+
+const dataValidationPerType = {
+  friendAcceptedRequest: ({ user }) => {
+    validations.pass('userId', user)
+  },
+  groupUpdate: ({ group, user, attribute, previousValue, newValue }) => {
+    validations.pass('groupId', group)
+    validations.pass('userId', user)
+    passFromFunction('attribute', attribute, _.isString)
+    Group.validations.pass(attribute, previousValue)
+    Group.validations.pass(attribute, newValue)
```

Comment:
those have already been validated by the group itself, when it was updated no? I dont think its the concern of the Notification model to know how is group updated (otherwise notification will end up knowing more models)

Reply:
here the Notification model doesn't know anything else than this attribute is known by the Group model, which should be able to confirm that it looks right before we sent it to the database

But I'm open to suggestions to replace it, including deleting it

Reply:
So in the end, every value in a model will end up being re-checked when a new notification is created.
What happens if this notification validation dont pass ? As far as i understand, the action which triggered the notification would be complete (a group action here) but the notification wont be created.
I think Notification model have to trust who is triggering it. What it must know though is that if value is a string in order to properly display it in the client.

I would delete those two lines.

Reply:
imo that's the role of the model to assert that what is put in the database conforms to certain expectations (expectations which are much needed to not make NoSQL databases more chaotic than they already are), and to throw errors at consumers that don't respect those expectations, allowing maintainers to correct those consumers' behavior: ideally those errors would be detected and fixed during development, but if happening in prod, it's preferable to throw an error at the closest point from the problem rather than letting through inconsistent data that would trigger errors down the road when expectations on which db queries build are not met

Also, those tests are very cheap

Reply:
> those have already been validated by the group itself, when it was updated no?

the notification model doesn't know that

Reply:
> it's preferable to throw an error at the closest point from the problem
Totally agree that's why the Group model has its own validation, and validated the data already
A model using another model's validation is opening a lot of unnecessary validations, i disagree on the cheapness, its mixing models, could become spaghettis code which is already the case in the controllers (circular dependency stuff) but model were pretty clean so far. Would be nice to keep it this way.

> the notification model doesn't know that
The notification model doesnt have to know anything else than how to produce a notification. Group validation is not the job of the notification model

Reply:
> The notification model doesnt have to know anything else than how to produce a notification. Group validation is not the job of the notification model

but this validation **is** intended to produce a notification, namely a group update notification, which is a sub-type of notification associated with the group model, just like the patch model is associated with the entity model, or the transaction model with the item model; calling elements of validation from that associated model makes explicit that association and doesn't chock me, au contraire

> A model using another model's validation is opening a lot of unnecessary validations
runtime assertions (just like all those `assert_.string`) are always unnecessary, rechecking the obvious, the already validated upstream, until they are not because they help you detect a bug before it makes it to the database: I like runtime assertions

I would thus still prefer to keep those, but wouldn't make this issue a blocking point, so if this is blocking for you, I can remove it (with a hint of grunting, but that would be quickly forgotten)

Reply:
after thinking about it, i think its necessary considering functional programming cannot ensure data consistency. In those cases, i feel that OOP is so much superior to keep validations at its place only.
I still dont feel good about mixing types of validations, but i guess i just have to swallow it and grumbling about FP